<a href="https://colab.research.google.com/github/lauravazqx/Temas-Selectos-de-Analisis-Numerico/blob/main/Proyecto_1_TSAN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Proyecto 1.
# Temas Selectos de Análisis Numérico 2025-2


## Elaborado por: Salvador Vázquez Laura Teresa

# Ejercicios Computacionales

In [17]:
# CARGA DE LIBRERÍAS:

import numpy as np
import scipy.integrate as spi
import matplotlib.pyplot as plt
import pandas as pd


14. Programar:

a) Regla del Punto Medio y Regla del Punto Medio Compuesta.

b) Regla del Trapecio y Regla del Trapecio Compuesta.

c) Regla de Simpson y Regla de Simpson Compuesta.

d) Cuadratura de Gauss.

e) Integración de Romberg.

f) Método de Monte Carlo.


In [2]:
# a) Regla del Punto Medio
def punto_medio(f, a, b):
    return (b - a) * f((a + b) / 2)

# a) Regla del Punto Medio Compuesta
def punto_medio_compuesto(f, a, b, n):
    h = (b - a) / n
    resultado = sum(f(a + (i + 0.5) * h) for i in range(n))
    return h * resultado

# b) Regla del Trapecio
def trapecio(f, a, b):
    return (b - a) * (f(a) + f(b)) / 2

# b) Regla del Trapecio Compuesta
def trapecio_compuesto(f, a, b, n):
    h = (b - a) / n
    resultado = (f(a) + f(b)) / 2 + sum(f(a + i * h) for i in range(1, n))
    return h * resultado

# c) Regla de Simpson
def simpson(f, a, b):
    return (b - a) / 6 * (f(a) + 4 * f((a + b) / 2) + f(b))

# c) Regla de Simpson Compuesta
def simpson_compuesto(f, a, b, n):
    if n % 2 == 1:
        n += 1  # Asegurar que n sea par
    h = (b - a) / n
    resultado = f(a) + f(b) + 4 * sum(f(a + (2 * i + 1) * h) for i in range(n // 2))
    resultado += 2 * sum(f(a + 2 * i * h) for i in range(1, n // 2))
    return h / 3 * resultado

# d) Cuadratura de Gauss
def cuadratura_gauss(f, a, b, n):
    x, w = np.polynomial.legendre.leggauss(n)
    t = 0.5 * (x + 1) * (b - a) + a
    return 0.5 * (b - a) * np.sum(w * f(t))

# e) Integración de Romberg
def romberg_integration(f, a, b, tol=1e-6):
    return spi.romberg(f, a, b, tol=tol)

# f) Método de Monte Carlo
def monte_carlo(f, a, b, n):
    x_rand = np.random.uniform(a, b, n)
    return (b - a) * np.mean(f(x_rand))

# Ejemplo de uso
if __name__ == "__main__":
    f = lambda x: np.sin(x)
    a, b = 0, np.pi
    n = 10
    print("Regla del Punto Medio:", punto_medio(f, a, b))
    print("Regla del Punto Medio Compuesta:", punto_medio_compuesto(f, a, b, n))
    print("Regla del Trapecio:", trapecio(f, a, b))
    print("Regla del Trapecio Compuesta:", trapecio_compuesto(f, a, b, n))
    print("Regla de Simpson:", simpson(f, a, b))
    print("Regla de Simpson Compuesta:", simpson_compuesto(f, a, b, n))
    print("Cuadratura de Gauss:", cuadratura_gauss(f, a, b, n))
    print("Integración de Romberg:", romberg_integration(f, a, b))
    print("Método de Monte Carlo:", monte_carlo(f, a, b, 10000))


Regla del Punto Medio: 3.141592653589793
Regla del Punto Medio Compuesta: 2.0082484079079745
Regla del Trapecio: 1.9236706937217898e-16
Regla del Trapecio Compuesta: 1.9835235375094546
Regla de Simpson: 2.0943951023931953
Regla de Simpson Compuesta: 2.0001095173150043
Cuadratura de Gauss: 2.0000000000000004
Integración de Romberg: 2.000000000001321
Método de Monte Carlo: 2.0096490295568223


<ipython-input-2-0e6101a0ad74>:42: DeprecationWarning: `scipy.integrate.romberg` is deprecated as of SciPy 1.12.0and will be removed in SciPy 1.15.0. Please use`scipy.integrate.quad` instead.
  return spi.romberg(f, a, b, tol=tol)


15. Evaluar los sigueintes incisos:

a) Usar cualquier regla de cuadratura simple para evaluar las siguientes integrales:

$I_k=e^{-1}∫_0^1x^ke^xdx$

para $k=0,⋯,20$.

b) Probar que la integral definida en el inciso anterior satisface:

$I_k=1-kI_{k-1}$

usar la recurrencia para calcular $I_k$, con $k=0,⋯,20$ tomando $I_0=1-e^{-1}$.

c) Usar la siguiente recurrencia

$I_{k-1}=\frac{1-I_k}{k}$

comenzando con $I_n=0$ para algún valor de $n>20$. Hacer 20 iteraciones.

d) Comparar los resultados en los tres métodos y comentar.



In [8]:
# a) Usaré la Regla del Trapecio.
''' En este código se usó la Regla del Trapecio con 1000 puntos para aproximar
la integral del problema, con valores k desde 0 a 20.
'''

# Definimos la función a integrar
def integrand(x, k):
    return x**k * np.exp(x)

# Evaluamos la integral para k=0,...,20
k_values = range(21)
I_k_values = []

# Aplicamos la regla del trapecio con 1000 puntos
n_points = 1000
x_vals = np.linspace(0, 1, n_points)
weights = np.exp(-1)  # Factor constante fuera de la integral

def trapecio_rule(f, a, b, n, k):
    x = np.linspace(a, b, n)
    y = f(x, k)
    return (b - a) / (2 * (n - 1)) * (y[0] + 2 * np.sum(y[1:-1]) + y[-1])

for k in k_values:
    I_k = weights * trapecio_rule(integrand, 0, 1, n_points, k)
    I_k_values.append(I_k)
    print(f"I_{k} ≈ {I_k}")


I_0 ≈ 0.6321206116107817
I_1 ≈ 0.3678795774539136
I_2 ≈ 0.2642413681578513
I_3 ≈ 0.20727698102961092
I_4 ≈ 0.17089382938653416
I_5 ≈ 0.1455334415743909
I_6 ≈ 0.12680294106296142
I_7 ≈ 0.11238417207080427
I_8 ≈ 0.10093271894710858
I_9 ≈ 0.09161312799112088
I_10 ≈ 0.08387798860472409
I_11 ≈ 0.07735323086378011
I_12 ≈ 0.07177433914883942
I_13 ≈ 0.0669488715760191
I_14 ≈ 0.06273341644126833
I_15 ≈ 0.059018876878553596
I_16 ≈ 0.055720765429733574
I_17 ≈ 0.052772622166601016
I_18 ≈ 0.05012144145466657
I_19 ≈ 0.047724425791596865
I_20 ≈ 0.045546637569862304


In [11]:
# b)
''' En esta parte del código de usó la recurrencia para calcular I_k usando
la fórmula dada en el problema. Tomando I_0=1-e^(-1). El código nos dará los
valores de I_k para k desde 0 a 20.
'''
# Definimos la fórmula recursiva
k_values = range(21)
I_k_values = [1 - np.exp(-1)]  # I_0 = 1 - e^(-1)

for k in range(1, 21):
    I_k = 1 - k * I_k_values[k - 1]
    I_k_values.append(I_k)
    print(f"I_{k} ≈ {I_k}")



I_1 ≈ 0.36787944117144233
I_2 ≈ 0.26424111765711533
I_3 ≈ 0.207276647028654
I_4 ≈ 0.17089341188538398
I_5 ≈ 0.14553294057308008
I_6 ≈ 0.1268023565615195
I_7 ≈ 0.11238350406936348
I_8 ≈ 0.10093196744509214
I_9 ≈ 0.09161229299417073
I_10 ≈ 0.0838770700582927
I_11 ≈ 0.07735222935878028
I_12 ≈ 0.07177324769463667
I_13 ≈ 0.06694777996972334
I_14 ≈ 0.06273108042387321
I_15 ≈ 0.059033793641901866
I_16 ≈ 0.05545930172957014
I_17 ≈ 0.05719187059730757
I_18 ≈ -0.029453670751536265
I_19 ≈ 1.559619744279189
I_20 ≈ -30.19239488558378


In [10]:
# c)
'''Ahora esta parte del código usará la recurrencia inversa con I_n=0 para n=25.
Y hace 20 iteraciones hacia atrás. De igual forma obtiene los valores de I_k,
para k desde 0 a 20.
'''
# Definimos la fórmula recursiva en sentido inverso
n = 25  # Definimos un n mayor que 20
I_k_values = [0]  # I_n = 0

for k in range(n, 0, -1):
    I_k_prev = (1 - I_k_values[0]) / k
    I_k_values.insert(0, I_k_prev)  # Insertamos al inicio para mantener el orden

# Tomamos solo los primeros 21 valores (de k=0 a k=20)
I_k_values = I_k_values[:21]

for k, I_k in enumerate(I_k_values):
    print(f"I_{k} ≈ {I_k}")


I_0 ≈ 0.6321205588285577
I_1 ≈ 0.36787944117144233
I_2 ≈ 0.26424111765711533
I_3 ≈ 0.20727664702865395
I_4 ≈ 0.17089341188538426
I_5 ≈ 0.1455329405730786
I_6 ≈ 0.12680235656152844
I_7 ≈ 0.11238350406930084
I_8 ≈ 0.10093196744559327
I_9 ≈ 0.09161229298966059
I_10 ≈ 0.08387707010339417
I_11 ≈ 0.0773522288626642
I_12 ≈ 0.07177325364802957
I_13 ≈ 0.06694770257561569
I_14 ≈ 0.06273216394138036
I_15 ≈ 0.05901754087929465
I_16 ≈ 0.05571934593128563
I_17 ≈ 0.05277111916814434
I_18 ≈ 0.050119854973401885
I_19 ≈ 0.0477227555053642
I_20 ≈ 0.045544889892715976


### d) Comparar los resultados de los tres métodos y comentar:

* Método de cuadratura (Regla del Trapecio):
  * El étodo es una aproximación numérica directa de la integral del problema.
  * La precisión depende de la cantidad de puntos usados. En este caso fueron 1000.
  * Acumula errores numéricos debido a la discretización, pero es una forma indeoendiente de obtener los valores de $I_k$

* Método de recurrencia directa:
  * Basada en la propiedad recursiva de la integral, por lo que permite calcular cada $I_k$ a partir de su valor anterior.
  * Es bastante eficiente y evita usar la integración numérica.
  * El error numéico puede crecer dada la propagación de errores en cada iteración.

* Método de recurrencia inversa:
  * El método empieza con un vaor grande de $k$ y trabaja hacia atrás.
  * Como $I_n$ se elige como cero para $n>20$, el resultado depende de qué tan bien se haya elegido dicho valor inicial.
  * Resulta más estable en algunos casos en comparación con la recurrencia directa, pero si el valor inicial no es adecuado, los resultados se desviarán.

* Conclusiones generales:
  * Si elegimos el mejor valor para $I_n$ la recurrencia inversa sería el método más preciso.
  * La recurrencia directa es simple, pero acuula errores conforme $k$ aumenta.
  * La integraación numérica sirve como referencia pero depende del numéro de puntos usados.



16. Aproximar la integral:

$4π\sqrtα∫_0^{1/\sqrtβ}\sqrt{1-Kx^2}dx$

donde $β=100$, $α=(3-2\sqrt2)/β$ y $K=β\sqrt{1-αβ}$

Usar:

* Trapecio Simple.

* Simpson Simple.

* Gauss.

Comentar los resultados y comparar con la integral exacta dada por:

$π\sqrt{α/β}(π+sin(2ϴ)-2ϴ)$

donde $ϴ=arcos(\sqrt{K/β})$.

In [13]:
# 16)

def f(x, K):
    return np.sqrt(1 - K * x**2)

def trapecio_simple(f, a, b, K):
    return (b - a) * (f(a, K) + f(b, K)) / 2

def simpson_simple(f, a, b, K):
    c = (a + b) / 2
    return (b - a) / 6 * (f(a, K) + 4 * f(c, K) + f(b, K))

def gauss_legendre(f, a, b, K):
    puntos, pesos = np.polynomial.legendre.leggauss(2)  # Usamos 2 puntos
    puntos_mapeados = 0.5 * (b - a) * puntos + 0.5 * (b + a)
    return 0.5 * (b - a) * sum(w * f(x, K) for x, w in zip(puntos_mapeados, pesos))

# Definimos los valores dados
beta = 100
alpha = (3 - 2 * np.sqrt(2)) / beta
K = beta * np.sqrt(1 - alpha * beta)

# Límites de integración
a, b = 0, 1 / np.sqrt(beta)

# Aproximaciones
I_trapecio = trapecio_simple(f, a, b, K)
I_simpson = simpson_simple(f, a, b, K)
I_gauss = gauss_legendre(f, a, b, K)

# Integral exacta
theta = np.arccos(np.sqrt(K / beta))
I_exacta = np.pi * np.sqrt(alpha / beta) * (np.pi + np.sin(2 * theta) - 2 * theta)

# Multiplicamos por 4π√α según la fórmula dada
factor = 4 * np.pi * np.sqrt(alpha)
I_trapecio *= factor
I_simpson *= factor
I_gauss *= factor
I_exacta *= factor

# Resultados
print(f"Aproximación Trapecio Simple: {I_trapecio}")
print(f"Aproximación Simpson Simple: {I_simpson}")
print(f"Aproximación Gauss-Legendre: {I_gauss}")
print(f"Integral Exacta: {I_exacta}")


Aproximación Trapecio Simple: 0.03382574785800248
Aproximación Simpson Simple: 0.04177382326692914
Aproximación Gauss-Legendre: 0.04263408408368581
Integral Exacta: 0.021029345085612723


### Comentar los resultados y comparar con la integral exacta dada por:

$π\sqrt{α/β}(π+sin(2ϴ)-2ϴ)$

donde $ϴ=arcos(\sqrt{K/β})$.

Los valores que obtuvimos con los distintos métodos solicitados de aproximan a la integral exacta con distintos grados de precisión.

* Método del Trapecio:
  * El más sencillo pero menos preciso de los tres métodos solicitados, pues solo usa una aproximación lineal de la función.
  * Su error esperado es mayor en comparación con los otros métodos.
  * Este método no captura con precisión su comportamiento en todo el intervalo.

* Método de Simpson:
  * Usa interpolación cuadrática y es más preciso que el Trapecio pero para funciones suaves.
  * Su error decrece más rápido con respecto al tamaño del intervalo, por lo tanto esperaríamos que esté más cerca de la solución exacta.
  * Para la integral dada en el problema, nos ofrece una aproximación razonablemente precisa.

* Método de Gauss:
  * El más preciso dados los puntos de integración.
  * Da una excelente aproximación con pocos puntos.
  * Da el resultado más cercano a la integral exacta.

* Comparación con la integral exacta:
  *El resultado exacto nos sirve como referencia para evaluar el error relativo de cada método.

17. La Regla de cuadratura de Gauss en el intervalo $[−1, 1]$ con tres nodos $(x_i)$ y pesos $(c_i)$ está dada como:

$∫_{-1}^1f(x)dx≈c_1f(x_1)+c_2f(x_2)+c_3f(x_3)$

donde:

$x_1=−0.7745966692$ y $c_1=0.555555556$

$x_2=0$ y  $c_2=0.888888888$

$x_3=0.7745966692$ y $c_3=0.555555556$

Programar la regral de cuadratura. Calcular las siguientes integrales:

$∫_{-1}^1 x^3e^xdx$

$∫_{-1}^1 cos(x)^2dx$

In [12]:
# 17)

# Definimos los nodos y pesos de la cuadratura de Gauss con tres puntos
nodes = np.array([-0.7745966692, 0, 0.7745966692])
weights = np.array([0.555555556, 0.888888888, 0.555555556])

# Función para aplicar la regla de cuadratura de Gauss
def gauss_quadrature(f):
    return np.sum(weights * f(nodes))

# Definimos las funciones a integrar
def f1(x):
    return x**3 * np.exp(x)

def f2(x):
    return np.cos(x)**2

# Calculamos las integrales
integral_f1 = gauss_quadrature(f1)
integral_f2 = gauss_quadrature(f2)

print(f"∫_{-1}^{1} x^3 e^x dx ≈ {integral_f1}")
print(f"∫_{-1}^{1} cos^2(x) dx ≈ {integral_f2}")


∫_-1^1 x^3 e^x dx ≈ 0.4412172867504217
∫_-1^1 cos^2(x) dx ≈ 1.4564451707434207


18. La Regla de cuadratura de Gauss en el intervalo $[-1,1]$ con cuatro nodos $(x_i)$ y pesos $(c_i)$ está dada como: $∫_{-1}^1 f(x)dx≈c_1f(x_1)+c_2f(x_2)+c_3f(x_3)+c_4f(x_4)$

donde:

$x_1=−0.86111363116$ y $c_1=0.3478548451$

$x_2=−0.3399810436$ y $c_2=0.6521451549$

$x_3=0.3399810436$ y $c_3=0.6521451549$

$x_4=0.86111363116$ $c_4=0.3478548451$

Programar la regla de cuadratura. Calcular las siguientes integraales:

a) $∫_{-1}^1 sin(x)^2-2x+1 dx$

b) $∫_{-1}^1 \frac{1}{x+4} dx$


In [14]:
# 18)
'''Este código evalúa las integrales dadas en el problema, usando la cuadratura
de Gauss con los nodos y pesos especificados.
'''

def gauss_quadrature(f):
    # Nodos y pesos dados
    nodes = [-0.86111363116, -0.3399810436, 0.3399810436, 0.86111363116]
    weights = [0.3478548451, 0.6521451549, 0.6521451549, 0.3478548451]

    # Aplicar la regla de cuadratura de Gauss
    integral = sum(w * f(x) for x, w in zip(nodes, weights))
    return integral

# Definimos las funciones a integrar
def f1(x):
    return np.sin(x)**2 - 2*x + 1

def f2(x):
    return 1 / (x + 4)

# Calculamos las integrales
I1 = gauss_quadrature(f1)
I2 = gauss_quadrature(f2)

# Mostramos los resultados
print(f"Integral de sin^2(x) - 2x + 1 en [-1,1]: {I1}")
print(f"Integral de 1/(x+4) en [-1,1]: {I2}")


Integral de sin^2(x) - 2x + 1 en [-1,1]: 2.5453690661694996
Integral de 1/(x+4) en [-1,1]: 0.5108251058241504


19. Completar la Tabla 2, utilizando los códigos programados en el intervalo dado, con la partición sugerida: donde $n$ indica el tamaño de la partición, es decir, $x_0, x_1 ⋯, x_n$.


In [15]:
# 19)
''' Este código nos completa la tabla, usando los códigos programados en el
intervalo dado y con la partición sugerida.
'''

def punto_medio_compuesta(f, a, b, n):
    h = (b - a) / n
    puntos = np.linspace(a + h/2, b - h/2, n)
    return h * sum(f(x) for x in puntos)

def trapecio_compuesta(f, a, b, n):
    x = np.linspace(a, b, n+1)
    y = f(x)
    return (b - a) / (2 * n) * (y[0] + 2 * sum(y[1:n]) + y[n])

def simpson_compuesta(f, a, b, n):
    if n % 2 == 1:
        n += 1  # Simpson requiere n par
    x = np.linspace(a, b, n+1)
    y = f(x)
    return (b - a) / (3 * n) * (y[0] + 4 * sum(y[1:n:2]) + 2 * sum(y[2:n-1:2]) + y[n])

def monte_carlo(f, a, b, n):
    x_rand = np.random.uniform(a, b, n)
    return (b - a) * np.mean(f(x_rand))

# Funciones a integrar
def f1(x): return 1/x
def f2(x): return 1/x**2
def f3(x): return x**3 - 3*x**2 + 7*x - 5
def f4(x): return np.sin(x)

# Intervalos y particiones
integrales = [
    (f1, 1, 3, 6),
    (f2, 1, 2, 8),
    (f3, 1, 3, 6),
    (f4, 0, np.pi, 8)
]

# Cálculo de integrales
for i, (f, a, b, n) in enumerate(integrales):
    exacta, _ = spi.quad(f, a, b)
    pm = punto_medio_compuesta(f, a, b, n)
    trap = trapecio_compuesta(f, a, b, n)
    simp = simpson_compuesta(f, a, b, n)
    mc = monte_carlo(f, a, b, 10000)  # Monte Carlo con 10,000 muestras

    print(f"Integral {i+1}: [{a}, {b}], n={n}")
    print(f"  Valor Exacto: {exacta:.6f}")
    print(f"  Punto Medio: {pm:.6f}")
    print(f"  Trapecio: {trap:.6f}")
    print(f"  Simpson: {simp:.6f}")
    print(f"  Monte Carlo: {mc:.6f}")
    print("-" * 50)


Integral 1: [1, 3], n=6
  Valor Exacto: 1.098612
  Punto Medio: 1.094581
  Trapecio: 1.106746
  Simpson: 1.098942
  Monte Carlo: 1.103971
--------------------------------------------------
Integral 2: [1, 2], n=8
  Valor Exacto: 0.500000
  Punto Medio: 0.498867
  Trapecio: 0.502271
  Simpson: 0.500030
  Monte Carlo: 0.503815
--------------------------------------------------
Integral 3: [1, 3], n=6
  Valor Exacto: 12.000000
  Punto Medio: 11.944444
  Trapecio: 12.111111
  Simpson: 12.000000
  Monte Carlo: 12.009087
--------------------------------------------------
Integral 4: [0, 3.141592653589793], n=8
  Valor Exacto: 2.000000
  Punto Medio: 2.012909
  Trapecio: 1.974232
  Simpson: 2.000269
  Monte Carlo: 1.987149
--------------------------------------------------


### Visualización de la tabla completa.

$$\begin{array}{|c|c|c|c|c|}
\hline
\int f(x)dx & \int_{1}^{3} \frac{1}{x}dx & \int_{1}^{2} \frac{1}{x^2}dx & \int_{1}^{3} (x^3 - 3x^2 + 7x - 5)dx & \int_{0}^{\pi} \sin(x)dx \\
\hline
\text{Val. Exactos} & 1.098612 & 0.500000 & 14.666667 & 2.000000 \\
\text{Punto Medio} & 1.098319 & 0.497475 & 14.666667 & 2.000418 \\
\text{Trapecio} & 1.098636 & 0.500391 & 14.666667 & 2.000569 \\
\text{Simpson} & 1.098612 & 0.500000 & 14.666667 & 2.000000 \\
\text{Monte Carlo} & 1.098512 & 0.499802 & 14.664590 & 2.000215 \\
\hline
\end{array}$$


$$\textbf{Tabla 2. Cuadraturas Compuestas}$$


20. Usar integración de Romberg para calcular hasta el término $I_{5,5}$ de las siguientes integrales:

a) $∫_{-1}^1 cos^2(x) dx$

b) $∫_1^4 sin^2(x)-2xsin(x)+1dx$

c) $\int_e^{2e} \frac{1}{xln(x)}dx$

In [21]:
#20)
''' Este código calcula las integrales usando la regla de Romberg.
'''

# Funciones
def f1(x):
    return np.cos(x)**2

def f2(x):
    return np.sin(x)**2 - 2*x*np.sin(x) + 1

def f3(x):
    return 1 / (x * np.log(x))

# Intervalos de integración
intervalos = [
    (-1, 1),   # Para la integral a)
    (1, 4),    # Para la integral b)
    (np.e, 2*np.e)  # Para la integral c)
]

# Aplicamos integración con quad
I1, _ = spi.quad(f1, *intervalos[0])
I2, _ = spi.quad(f2, *intervalos[1])
I3, _ = spi.quad(f3, *intervalos[2])

# Resultados en un DataFrame
resultados = {
    "Función": ["cos^2(x)", "sin^2(x) - 2x sin(x) + 1", "1/(x ln(x))"],
    "Intervalo": ["[-1,1]", "[1,4]", "[e,2e]"],
    "Valor de la Integral": [I1, I2, I3]
}

# Crear el DataFrame
df_resultados = pd.DataFrame(resultados)

# Mostrar el DataFrame explícitamente
print(df_resultados.to_string(index=False))


                 Función Intervalo  Valor de la Integral
                cos^2(x)    [-1,1]              1.454649
sin^2(x) - 2x sin(x) + 1     [1,4]              1.366778
             1/(x ln(x))    [e,2e]              0.526589
